consensus/src/aggregation/engine.rs

```diff
@@ -0,0 +1,799 @@
+            return Err(Error::UnknownShare(self.epoch));
+        };
+        let ack = acks
+            .get(&self.epoch)
+            .and_then(|acks| acks.get(&share.index).cloned());
+        let ack = match ack {
+            Some(ack) => ack,
+            None => self.sign_ack(index, *digest).await?,
+        };
+
+        // Reinsert the index with a new deadline
+        self.set_rebroadcast_deadline(index);
+
+        // Broadcast the ack to all peers
+        self.broadcast(ack, sender).await
+    }
+
+    // ---------- Validation ----------
+
+    /// Takes a raw ack (from sender) from the p2p network and validates it.
+    ///
+    /// Returns the chunk, epoch, and partial signature if the ack is valid.
+    /// Returns an error if the ack is invalid.
+    fn validate_ack(&self, ack: &Ack<V, D>, sender: &P) -> Result<(), Error> {
+        // Validate epoch
+        {
+            let (eb_lo, eb_hi) = self.epoch_bounds;
+            let bound_lo = self.epoch.saturating_sub(eb_lo);
+            let bound_hi = self.epoch.saturating_add(eb_hi);
+            if ack.epoch < bound_lo || ack.epoch > bound_hi {
+                return Err(Error::AckEpochOutsideBounds(ack.epoch, bound_lo, bound_hi));
+            }
+        }
+
+        // Validate sender
+        let Some(sig_index) = self.validators.is_participant(ack.epoch, sender) else {
+            return Err(Error::UnknownValidator(ack.epoch, sender.to_string()));
+        };
+        if sig_index != ack.signature.index {
+            return Err(Error::PeerMismatch);
+        }
+
+        // Validate height
+        if ack.item.index < self.tip {
+            return Err(Error::AckThresholded(ack.item.index));
+        }
+        if ack.item.index >= self.tip + self.window {
+            return Err(Error::AckIndex(ack.item.index));
+        }
+
+        // Validate that we don't already have the ack
+        if self.confirmed.contains_key(&ack.item.index) {
+            return Err(Error::AckThresholded(ack.item.index));
+        }
+        let have_ack = match self.pending.get(&ack.item.index) {
+            None => false,
+            Some(Pending::Unverified(epoch_map)) => epoch_map
+                .get(&ack.epoch)
+                .is_some_and(|acks| acks.contains_key(&ack.signature.index)),
+            Some(Pending::Verified(_, epoch_map)) => epoch_map
+                .get(&ack.epoch)
+                .is_some_and(|acks| acks.contains_key(&ack.signature.index)),
+        };
+        if have_ack {
+            return Err(Error::AckDuplicate(sender.to_string(), ack.item.index));
+        }
+
+        // Validate partial signature
+        if !ack.verify(&self.namespace, self.validators.identity()) {
+            return Err(Error::InvalidAckSignature);
+        }
+
+        Ok(())
+    }
+
+    // ---------- Helpers ----------
+
+    /// Sets `index` as pending and requests the digest from the automaton.
+    fn get_digest(&mut self, index: Index) {
```

Comment:
nit: refer to this as an object (e.g. `get_item`) rather than by its type (just the means by which it is represented).

Reply:
This is fine for now 👍